### PR TITLE
Add Kosaraju's Algorithm for finding Strongly Connected Components in a directed graph.

### PR DESCRIPTION
**Pull request by sinhatushar**
This implementation runs faster as compared to the current implementation in src/connectivity.jl.
Some benchmarks on SNAPDatasets :
```
julia> eg = SimpleDiGraph{Int32}(loadsnap(:ego_twitter_d));

julia> @benchmark strongly_connected_components_kosaraju(eg)
BenchmarkTools.Trial: 
  memory estimate:  6.08 MiB
  allocs estimate:  49450
  --------------
  minimum time:     35.088 ms (0.00% GC)
  median time:      37.437 ms (0.00% GC)
  mean time:        37.453 ms (1.64% GC)
  maximum time:     41.985 ms (4.95% GC)
  --------------
  samples:          134
  evals/sample:     1

julia> @benchmark strongly_connected_components(eg)
BenchmarkTools.Trial: 
  memory estimate:  2.30 GiB
  allocs estimate:  36776
  --------------
  minimum time:     92.271 ms (57.10% GC)
  median time:      98.682 ms (58.91% GC)
  mean time:        106.764 ms (61.84% GC)
  maximum time:     196.805 ms (78.53% GC)
  --------------
  samples:          48
  evals/sample:     1
```
```
julia> ss = SimpleDiGraph{Int32}(loadsnap(:soc_slashdot0902_d));

julia> @benchmark strongly_connected_components(ss)
BenchmarkTools.Trial: 
  memory estimate:  1.70 GiB
  allocs estimate:  31705
  --------------
  minimum time:     86.266 ms (44.07% GC)
  median time:      90.336 ms (46.22% GC)
  mean time:        98.146 ms (49.64% GC)
  maximum time:     181.922 ms (73.07% GC)
  --------------
  samples:          51
  evals/sample:     1

julia> @benchmark strongly_connected_components_kosaraju(ss)
BenchmarkTools.Trial: 
  memory estimate:  5.34 MiB
  allocs estimate:  42541
  --------------
  minimum time:     42.838 ms (0.00% GC)
  median time:      44.965 ms (0.00% GC)
  mean time:        45.110 ms (1.03% GC)
  maximum time:     49.870 ms (3.12% GC)
  --------------
  samples:          111
  evals/sample:     1
```

### CODE REVIEW

**Comment by sinhatushar** [comment]
Also, the current implementation uses an enormous amount of memory as compared to this implementation.

File: src/connectivity.jl [hunk 1]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+               color[w] = 1
+           else
+               push!(order, u)  #Push back in vector to store the order in which the traversal finishes(Reverse Topological Sort)
+               color[u] = 2
+               pop!(dfs_stack)    
+           end
+       end
+   end
+    
+   reverse!(g)   # Reverse the graph (Transpose of the graph)
```
Comment (simonschoelly) [inline src/connectivity.jl]: We cannot use this trick with reversing twice here as this is not threadsafe. Also, this is only really refficient for graphs such `SimpleDiGraph, but will not be efficient for `SimpleWeightedDiGraph`.

File: src/connectivity.jl [hunk 2]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+ [11, 10]    
+ [2, 3, 4, 1]
+ [6, 7, 5]   
+ [9, 8]      
+
+```
+"""
+
+function strongly_connected_components_kosaraju end
+@traitfn function strongly_connected_components_kosaraju(g::AG::IsDirected) where {T, AG <: AbstractGraph{T}}
```
Comment (simonschoelly) [inline src/connectivity.jl]: `T` should be restricted to `Integer` here.

File: src/connectivity.jl [hunk 3]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+       component=Vector{T}()   # Vector used to store the vertices of one component temporarily
+       
+       # Start dfs from v
+       dfs_stack = Vector{T}([v])   # Stack used for dfs. Also push v to the stack
+       
+       while !isempty(dfs_stack)
+           u = dfs_stack[end]
+           w = 0
+       
+           for n in outneighbors(g, u)
```
Comment (simonschoelly) [inline src/connectivity.jl]: If you would replace `outneighbors` by `inneighbors` here, maybe it would be possible to avoid having to reverse the graph?

File: src/connectivity.jl [hunk 4]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
 
+"""
+    strongly_connected_components_kosaraju(g)
+
+Compute the strongly connected components of a directed graph `g` using Kosaraju's Algorithm. 
+(https://en.wikipedia.org/wiki/Kosaraju%27s_algorithm).
+
+Return an array of arrays, each of which is the entire connected component.
+
+### Performance
```
Comment (simonschoelly) [inline src/connectivity.jl]: Stating time and space complexity here is good,  but I think we should put the more detailed information in a comment somewhere else, as this is not relevant information when someone wants to quickly look up, what this function does.

File: src/connectivity.jl [hunk 5]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+This algorithm requires 2 depth first search passes. Also the algorithm reverses the graph once after one dfs.
+At the end, the graph is reversed again to restore the original graph.
+Also, an iterative version of dfs is written so that this graph doesn't show "stackoverflow" when run on large 
+graphs.
+
+Space Complexity : O(|V|) 
+This memory is used for vectors named color, original, component and components. This space complexity 
+excludes the space required to store the graph.
+
+### Examples
```
Comment (simonschoelly) [inline src/connectivity.jl]: It is good that you are putting examples here, but try to shorten this list.

Also instead of using `add_edge` in a loop you can  just write `g = SimpleDiGraph(Edge.(edgelist))`.

File: src/connectivity.jl [hunk 6]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+       color[v] = 1
+       
+       # Start dfs from v
+       dfs_stack = Vector{T}([v])   # Stack used for dfs. Also push v to the stack
+       
+       while !isempty(dfs_stack)
+           u = dfs_stack[end]
+           w = 0
+       
+           for n in outneighbors(g, u)
```
Comment (simonschoelly) [inline src/connectivity.jl]: Don't use `n` for a vertex, this will confuse people, as most of the time, they expect `n` to be a quantity. What about `u_neigh` or `u_neighbor`?

File: src/connectivity.jl [hunk 7]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+       
+       color[v] != 0  && continue  
+       color[v] = 1
+       
+       # Start dfs from v
+       dfs_stack = Vector{T}([v])   # Stack used for dfs. Also push v to the stack
+       
+       while !isempty(dfs_stack)
+           u = dfs_stack[end]
+           w = 0
```
Comment (simonschoelly) [inline src/connectivity.jl]: Either you write here `w = zero(T)`  or you replace later down the line `w = n` by `w = Int(n)`. Otherwise you will have some issues with type stability when `T` is not `Int`, as then the type of `w` will change at some point, which is bad for performance.

File: src/connectivity.jl [hunk 8]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+```
+"""
+
+function strongly_connected_components_kosaraju end
+@traitfn function strongly_connected_components_kosaraju(g::AG::IsDirected) where {T, AG <: AbstractGraph{T}}
+       
+   nvg = nv(g)    
+
+   components = Vector{Vector{T}}()    # Maintains a list of strongly connected components
+   sizehint!(components, nvg)
```
Comment (simonschoelly) [inline src/connectivity.jl]: I am not sure that this `sizehint` here is very wise. The problem is, that `sizehint!` usually does allocate a vector of the required size in the background. If we only have a few strongly connected components, this allocates more memory than necessary.

File: test/connectivity.jl [hunk 9]
```diff
@@ -111,12 +127,20 @@
...
       @test sort(scc[1]) == [7, 8, 9, 10, 11, 12]
       @test sort(scc[2]) == [3, 6]
       @test sort(scc[3]) == [2, 4, 5]
       @test scc[4] == [1]
+        
+      @test length(scc_k) == 4
+      @test sort(scc_k[1]) ==  [1] 
+      @test sort(scc_k[2]) ==  [2, 4 ,5] 
+      @test sort(scc_k[3]) ==  [3, 6]  
+      @test scc_k[4] ==  [11, 12, 10, 9, 7, 8]
```
Comment (simonschoelly) [inline test/connectivity.jl]: You should also put `sort` here. The reason that `sort`  was not used for `scc` was, that that component only had one element.

File: src/connectivity.jl [hunk 10]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+   reverse!(g)   # Reverse the graph (Transpose of the graph) 
+    
+   for i = 1:nvg
+        color[i] = 0    # Marking all the vertices from 1 to n as unvisited for dfs2
+   end
+   
+   # dfs2
+   for i in 1:nvg
+       
+       v = order[nvg-i+1]   # Reading the order vector in the decreasing order of finish time
```
Comment (simonschoelly) [inline src/connectivity.jl]: Maybe replace that by `order[end  - i + 1]`, sound a little bit more intuitive to me.

File: test/connectivity.jl (hunk 9 above)
Comment (simonschoelly) [reply test/connectivity.jl]: Also, does kosaraju guarantee that the components are sorted in increasing order of length? Otherwise we should sort the components by length first, before doing these tests.

File: src/connectivity.jl [hunk 11]
```diff
@@ -283,6 +283,204 @@ function strongly_connected_components end
...
+   color = zeros(UInt8, nvg)       # Vector used as for marking the colors during dfs
+   
+   # dfs1
+   for v in vertices(g)
+       
+       color[v] != 0  && continue  
+       color[v] = 1
+       
+       # Start dfs from v
+       dfs_stack = Vector{T}([v])   # Stack used for dfs. Also push v to the stack
```
Comment (simonschoelly) [inline src/connectivity.jl]: You should replace this by `dfs_stack  = T[v]`. The problem is, that allocating vectors is quite expensive and when you write `Vector{T}([v])` you first allocate a vector [v]  and then convert it to `T[v]`.

**Comment by sinhatushar** [comment]
> Your code does not look bad already, although I have not had time to completely understand it yet. The most important thing to fix, is the one with `reverse!` and then some smaller issues.

**Comment by sinhatushar** [comment]
I closed this by mistake. Please ignore.

File: src/connectivity.jl (hunk 1 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Ok, instead using `inneighbors` .

File: src/connectivity.jl (hunk 2 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed.

File: src/connectivity.jl (hunk 3 above)
Comment (sinhatushar) [reply src/connectivity.jl]: That was really smart @simonschoelly  _/\_. Using `inneighbors` works well and I don't know why I didn't think of this.

File: src/connectivity.jl [hunk 12]
```diff
@@ -284,6 +284,162 @@ function strongly_connected_components end
...
+   color = zeros(UInt8, nvg)       # Vector used as for marking the colors during dfs
+   
+   # dfs1
+   for v in vertices(g)
+       
+       color[v] != 0  && continue  
+       color[v] = 1
+       
+       # Start dfs from v
+       dfs_stack = T[v]   # Stack used for dfs. Also push v to the stack
```
Comment (simonschoelly) [inline src/connectivity.jl]: Put this line outside of the loop and then just use `empty! ` and `push! ` to insert a single element. This way, we avoid having to allocate memory twice.

File: src/connectivity.jl [hunk 13]
```diff
@@ -284,6 +284,162 @@ function strongly_connected_components end
...
+   for i in 1:nvg
+    
+       v = order[end-i+1]   # Reading the order vector in the decreasing order of finish time
+       color[v] != 0  && continue  
+       color[v] = 1
+       
+       component=Vector{T}()   # Vector used to store the vertices of one component temporarily
+       
+       # Start dfs from v
+       dfs_stack = T[v]   # Stack used for dfs. Also push v to the stack
```
Comment (simonschoelly) [inline src/connectivity.jl]: Also put this outside of the loop.

File: src/connectivity.jl (hunk 4 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Shortened the information.

File: src/connectivity.jl (hunk 5 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Making use of `g = SimpleDiGraph(Edge.(edgelist))` now.

File: src/connectivity.jl (hunk 6 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Chnaged `n` to `u_neighbor` .

File: src/connectivity.jl (hunk 8 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Removed `sizehint` .

File: test/connectivity.jl (hunk 9 above)
Comment (sinhatushar) [reply test/connectivity.jl]: Rewritten to  `sort(scc_k[4])` . No, neither Tarjan nor Kosaraju guarantee that components are sorted in increasing order of length. But why do we need to sort by length first ? The current written tests don't do this.

File: src/connectivity.jl (hunk 10 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed to `order[end-i+1]` .

File: src/connectivity.jl (hunk 11 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed to `dfs_stack = T[v]` .

File: src/connectivity.jl (hunk 7 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed to `w = zero(T)` .

File: src/connectivity.jl [hunk 14]
```diff
@@ -284,6 +284,162 @@ function strongly_connected_components end
...
+
+   components = Vector{Vector{T}}()    # Maintains a list of strongly connected components
+   
+   order = Vector{T}()         # Vector which will store the order in which vertices are visited
+   sizehint!(order, nvg)
+   
+   color = zeros(UInt8, nvg)       # Vector used as for marking the colors during dfs
+   
+   # dfs1
+   for v in vertices(g)
```
Comment (simonschoelly) [inline src/connectivity.jl]: Your loops have a lot of array access inside. For performance reasons, you should put `@inbounds` in front of the loop or do that individually for every array access.

**Comment by simonschoelly** [comment]
I think you are not testing for graphs with self  loops yet, so you should also do that. And also add a test for the special cases, where the graph has zero or one vertex.

File: src/connectivity.jl (hunk 12 above)
Comment (sinhatushar) [reply src/connectivity.jl]: We don't need to use `empty!` because the `while` loop breaks only when dfs_stack becomes empty.

Comment (sinhatushar) [reply src/connectivity.jl]: I have inserted `dfs_stack=T[]` at the beginning.

File: src/connectivity.jl (hunk 13 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Done.

**Comment by sinhatushar** [comment]
@simonschoelly , the above 2 type of tests are not there for current implementation also. I will write these 2 tests for that implementation as well.

File: src/connectivity.jl (hunk 14 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Ok, writing `@inbounds for v in bertices(g)`

**Comment by simonschoelly** [comment]
That is a good idea. It looks you have fixed everything else then. Can you rerun the benchmark? I am curious about that.

**Comment by sinhatushar** [comment]
Benchmark after making all the changes as suggested by @simonschoelly :
`
julia> eg = SimpleDiGraph{Int32}(loadsnap(:ego_twitter_d));

julia> @benchmark strongly_connected_components_kosaraju(eg)
BenchmarkTools.Trial: 
  memory estimate:  3.20 MiB
  allocs estimate:  24577
  --------------
  minimum time:     31.650 ms (0.00% GC)
  median time:      35.501 ms (0.00% GC)
  mean time:        36.798 ms (1.96% GC)
  maximum time:     91.619 ms (60.92% GC)
  --------------
  samples:          136
  evals/sample:     1

julia> @benchmark strongly_connected_components(eg)
BenchmarkTools.Trial: 
  memory estimate:  2.30 GiB
  allocs estimate:  36776
  --------------
  minimum time:     62.127 ms (37.91% GC)
  median time:      87.655 ms (46.96% GC)
  mean time:        104.415 ms (50.79% GC)
  maximum time:     251.884 ms (73.07% GC)
  --------------
  samples:          48
  evals/sample:     1
`

`
julia> ss = SimpleDiGraph{Int32}(loadsnap(:soc_slashdot0902_d));

julia> @benchmark strongly_connected_components_kosaraju(ss)
BenchmarkTools.Trial: 
  memory estimate:  2.90 MiB
  allocs estimate:  21169
  --------------
  minimum time:     38.419 ms (0.00% GC)
  median time:      40.337 ms (0.00% GC)
  mean time:        41.182 ms (1.62% GC)
  maximum time:     89.278 ms (56.63% GC)
  --------------
  samples:          122
  evals/sample:     1

julia> @benchmark strongly_connected_components(ss)
BenchmarkTools.Trial: 
  memory estimate:  1.70 GiB
  allocs estimate:  31705
  --------------
  minimum time:     125.253 ms (33.85% GC)
  median time:      135.431 ms (32.82% GC)
  mean time:        147.809 ms (37.98% GC)
  maximum time:     258.369 ms (62.59% GC)
  --------------
  samples:          34
  evals/sample:     1
`

**Comment by simonschoelly** [comment]
I repeated your benchmark on my computer, and interestingly, the first graph has the same ratio, but for the second it is only twice as fast. No idea why that is.
```
julia> eg = squash(loadsnap(:ego_twitter_d))
{81306, 1768149} directed simple UInt32 graph

julia> @btime strongly_connected_components_kosaraju($eg);
  28.576 ms (24577 allocations: 3.20 MiB)

julia> @btime strongly_connected_components($eg);
  63.558 ms (36776 allocations: 2.30 GiB)

julia> ss = squash(loadsnap(:soc_slashdot0902_d))
{82168, 948464} directed simple UInt32 graph

julia> @btime strongly_connected_components_kosaraju($ss);
  32.261 ms (21169 allocations: 2.90 MiB)

julia> @btime strongly_connected_components($ss);
  65.636 ms (31705 allocations: 1.70 GiB)
```

I am also under the impression, that the two algorithms perform nearly equally fast, when there is only one connected component.

Anyway, this shows us, thath Tarjan, which should in theory be faster, does not perform as fast as it should. Benchmarking shows us, that there happen a lot of allocations in tarjan,  so there is probably some room for improvements.

@sbromberger Fromy my side, this can be merged.

**Comment by sinhatushar** [comment]
I also observed a weird behaviour during benchmarking. Most of the times, Kosaraju will be 3-4 times faster but  sometimes only twice as fast (although it rarely happens). Now I am working on Tarjan's algorithm to see if there is any scope of improvement. The enormous amount of memory which Tarjan algortihm is needing is a bit weird.

**Comment by sinhatushar** [comment]
@sbromberger , waiting for you to merge this if you find this PR suitable. @simonschoelly  has approved these changes.

File: src/connectivity.jl [hunk 15]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+
+julia> strongly_connected_components_kosaraju(g)
+4-element Array{Array{Int64,1},1}:
+ [11, 10]    
+ [2, 3, 4, 1]
+ [6, 7, 5]   
+ [9, 8]      
+
+```
+"""
```
Comment (sbromberger) [inline src/connectivity.jl]: This is a very nice docstring :)

File: src/connectivity.jl [hunk 16]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+   nvg = nv(g)    
+
+   components = Vector{Vector{T}}()    # Maintains a list of strongly connected components
+   
+   order = Vector{T}()         # Vector which will store the order in which vertices are visited
+   sizehint!(order, nvg)
+   
+   color = zeros(UInt8, nvg)       # Vector used as for marking the colors during dfs
+   
+   dfs_stack = T[]   # Stack used for dfs
```
Comment (sbromberger) [inline src/connectivity.jl]: LG convention is to use the full declaration: `Vector{T}()`.

File: src/connectivity.jl [hunk 17]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+    
+   # dfs1
+   @inbounds for v in vertices(g)
+
```
Comment (sbromberger) [inline src/connectivity.jl]: In general, we don't need this much whitespace :)

File: src/connectivity.jl [hunk 18]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+   dfs_stack = T[]   # Stack used for dfs
+    
+   # dfs1
+   @inbounds for v in vertices(g)
+       
+       color[v] != 0  && continue  
+       color[v] = 1
+       
+       # Start dfs from v
+       push!(dfs_stack,v)   # Push v to the stack
```
Comment (sbromberger) [inline src/connectivity.jl]: space between arguments here.

File: src/connectivity.jl [hunk 19]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+           else
+               push!(order, u)  #Push back in vector to store the order in which the traversal finishes(Reverse Topological Sort)
+               color[u] = 2
+               pop!(dfs_stack)    
+           end
+       end
+   end
+    
+    
+   @inbounds for i = 1:nvg
```
Comment (sbromberger) [inline src/connectivity.jl]: If we're using `i` as an index, then this is ok. If `i` is supposed to represent vertices, then `for i in vertices(g)` is more appropriate. Your call.

File: src/connectivity.jl [hunk 20]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+                   break
+               end
+           end
+           
+           if w != 0
+               push!(dfs_stack, w)
+               color[w] = 1
+           else
+               color[u] = 2
+               push!(component,u)   # Push u to the vector component
```
Comment (sbromberger) [inline src/connectivity.jl]: space between arguments here.

File: src/connectivity.jl [hunk 21]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+               pop!(dfs_stack)    
+           end
+       end
+       
+       push!(components,component)
```
Comment (sbromberger) [inline src/connectivity.jl]: spacing

File: src/connectivity.jl [hunk 22]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+   end
+ 
+
```
Comment (sbromberger) [inline src/connectivity.jl]: whitespace

File: src/connectivity.jl [hunk 23]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+       end
+       
+       push!(components,component)
+   end
+ 
+    
+   return components
+end
+
+
```
Comment (sbromberger) [inline src/connectivity.jl]: whitespace

File: test/connectivity.jl [hunk 24]
```diff
@@ -75,31 +77,84 @@
       scc_as_subgraphs = map(i -> graph[i], scc)
       return all(is_strongly_connected, scc_as_subgraphs)
     end
+    
+    function scc_k_ok(graph)
+      """Check that all SCC really are strongly connected"""
```
Comment (sbromberger) [inline test/connectivity.jl]: this should be a comment, not a docstring (looks sort of like python :) )

File: src/connectivity.jl [hunk 25]
```diff
@@ -284,6 +284,164 @@ function strongly_connected_components end
...
+   @inbounds for i in 1:nvg
+    
+       v = order[end-i+1]   # Reading the order vector in the decreasing order of finish time
+       color[v] != 0  && continue  
+       color[v] = 1
+       
+       component=Vector{T}()   # Vector used to store the vertices of one component temporarily
+       
+       # Start dfs from v
+       push!(dfs_stack,v)   # Push v to the stack
```
Comment (sbromberger) [inline src/connectivity.jl]: space between arguments

Comment (sinhatushar) [reply src/connectivity.jl]: Changed.

File: test/connectivity.jl (hunk 24 above)
Comment (sinhatushar) [reply test/connectivity.jl]: Changed.

File: src/connectivity.jl (hunk 23 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Removed.

File: src/connectivity.jl (hunk 22 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Removed.

File: src/connectivity.jl (hunk 21 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed.

File: src/connectivity.jl (hunk 20 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed.

File: src/connectivity.jl (hunk 19 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed.

File: src/connectivity.jl (hunk 18 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed.

File: src/connectivity.jl (hunk 17 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Removed.

File: src/connectivity.jl (hunk 16 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Changed.

File: src/connectivity.jl [hunk 26]
```diff
@@ -284,6 +284,158 @@ function strongly_connected_components end
...
+               color[w] = 1
+           else
+               push!(order, u)  #Push back in vector to store the order in which the traversal finishes(Reverse Topological Sort)
+               color[u] = 2
+               pop!(dfs_stack)    
+           end
+       end
+   end
+    
+   @inbounds for i in vertices(g)
```
Comment (eulerkochy) [inline src/connectivity.jl]:
```suggestion
   @inbounds for i in 1:nvg
```

File: src/connectivity.jl [hunk 27]
```diff
@@ -284,6 +284,158 @@ function strongly_connected_components end
...
+   
+   order = Vector{T}()         # Vector which will store the order in which vertices are visited
+   sizehint!(order, nvg)    
+   
+   color = zeros(UInt8, nvg)       # Vector used as for marking the colors during dfs
+   
+   dfs_stack = Vector{T}()   # Stack used for dfs
+    
+   # dfs1
+   @inbounds for v in vertices(g)
```
[the remark this answers is not in the document]
Comment (sinhatushar) [reply src/connectivity.jl]: @eulerkochy , I'm using `for v in vertices(g)` as suggested by @sbromberger .

File: src/connectivity.jl (hunk 26 above)
Comment (sinhatushar) [reply src/connectivity.jl]: Read above.

File: src/connectivity.jl (hunk 27 above)
[the remark this answers is not in the document]
Comment (eulerkochy) [reply src/connectivity.jl]: Sure, then!
